### ISSUE TITLE
WKTReader always throws InvalidOperationException

### ISSUE THREAD

**Issue by NightOwl888** [NetTopologySuite/NetTopologySuite#189]
The line:

```
new WKTReader(m_ctx.GeometryFactory).Read(str);
```

Always throws an InvalidOpertationException:

```
Test Name:	Spatial4n.Core.IO.NtsWKTReaderShapeParserTest.WktGeoPt
Test FullName:	Spatial4n.Core.IO.NtsWKTReaderShapeParserTest.WktGeoPt
Test Source:	F:\Projects\Spatial4n\Spatial4n.Tests\IO\NtsWKTReaderShapeParserTest.cs : line 44
Test Outcome:	Failed
Test Duration:	0:00:00.004

Result StackTrace:	
at Spatial4n.Core.IO.Nts.NtsWKTReaderShapeParser.ParseIfSupported(String str, WKTReader reader) in F:\Projects\Spatial4n\Spatial4n.Core.NTS\IO\Nts\NtsWKTReaderShapeParser.cs:line 91
   at Spatial4n.Core.IO.Nts.NtsWKTReaderShapeParser.ParseIfSupported(String wktString) in F:\Projects\Spatial4n\Spatial4n.Core.NTS\IO\Nts\NtsWKTReaderShapeParser.cs:line 50
   at Spatial4n.Core.IO.WktShapeParser.Parse(String wktString) in F:\Projects\Spatial4n\Spatial4n.Core\IO\WktShapeParser.cs:line 90
   at Spatial4n.Core.Context.SpatialContext.ReadShapeFromWkt(String wkt) in F:\Projects\Spatial4n\Spatial4n.Core\Context\SpatialContext.cs:line 382
   at Spatial4n.Core.Context.SpatialContext.ReadShape(String value) in F:\Projects\Spatial4n\Spatial4n.Core\Context\SpatialContext.cs:line 406
   at Spatial4n.Core.IO.NtsWKTReaderShapeParserTest.WktGeoPt() in F:\Projects\Spatial4n\Spatial4n.Tests\IO\NtsWKTReaderShapeParserTest.cs:line 46
----- Inner Stack Trace -----
   at GeoAPI.GeometryServiceProvider.ReflectInstance()
   at GeoAPI.GeometryServiceProvider.get_Instance()
   at NetTopologySuite.IO.WKTReader.ReadGeometryTaggedText(IEnumerator`1 tokens)
   at NetTopologySuite.IO.WKTReader.Read(TextReader reader)
   at NetTopologySuite.IO.WKTReader.Read(String wellKnownText)
   at Spatial4n.Core.IO.Nts.NtsWKTReaderShapeParser.ParseIfSupported(String str, WKTReader reader) in F:\Projects\Spatial4n\Spatial4n.Core.NTS\IO\Nts\NtsWKTReaderShapeParser.cs:line 63
Result Message:	
Spatial4n.Core.Exceptions.InvalidShapeException : error reading WKT: System.InvalidOperationException: Cannot use GeometryServiceProvider without an assigned IGeometryServices class
   at GeoAPI.GeometryServiceProvider.ReflectInstance()
   at GeoAPI.GeometryServiceProvider.get_Instance()
   at NetTopologySuite.IO.WKTReader.ReadGeometryTaggedText(IEnumerator`1 tokens)
   at NetTopologySuite.IO.WKTReader.Read(TextReader reader)
   at NetTopologySuite.IO.WKTReader.Read(String wellKnownText)
   at Spatial4n.Core.IO.Nts.NtsWKTReaderShapeParser.ParseIfSupported(String str, WKTReader reader) in F:\Projects\Spatial4n\Spatial4n.Core.NTS\IO\Nts\NtsWKTReaderShapeParser.cs:line 63
---- System.InvalidOperationException : Cannot use GeometryServiceProvider without an assigned IGeometryServices class
```

Analyzing the source, I can see that it fails on [this line](https://github.com/NetTopologySuite/NetTopologySuite/blob/af938c521caf2c5bd3ce122ecc9b73b6bf10dad0/NetTopologySuite/IO/WKTReader.cs#L445-L446) because apparently, the [HAS_SYSTEM_APPDOMAIN_GETASSEMBLIES symbol](https://github.com/NetTopologySuite/GeoAPI/blob/af655b3440ac76117ea4f23845a6c1d3dba96de1/src/GeoAPI/GeometryServiceProvider.cs#L75-L99) was not used when GeoAPI 1.7.4 (and 1.7.5-pre4) was built.

I realize there is an issue here with scanning assemblies on .NET Standard. Do note that it is possible on .NET Standard 1.5 by using `Microsoft.Extensions.DependencyModel` version 1.0.0-rc2-final ([example](https://github.com/apache/lucenenet/blob/29cbc138b5096a0063d3bfa89a26e306c7f7b710/src/Lucene.Net/Support/AssemblyUtils.cs#L29-L35)), but it looks like they only support .NET Standard 1.6 on the 1.0.0 release.

You can see the failure by cloning [this branch](https://github.com/NightOwl888/Spatial4n/tree/nts-wktreader-bug) of spatial4n and running the `Spatial4n.Core.IO.NtsWKTReaderShapeParserTest.WktGeoPt` test. To change the target framework, you can enable [this line](https://github.com/NightOwl888/Spatial4n/blob/nts-wktreader-bug/Spatial4n.Tests/Spatial4n.Tests.csproj#L8) and set it to one of the values in `<TargetFrameworks>`.

Looking at the source, there doesn't appear to be a way to inject an alternative implementation of `IGeometryServices` (which would be the ideal fix). You should really consider adding a seam through the use of an [abstract factory](http://blog.ploeh.dk/2014/05/19/di-friendly-framework/) so the end user can supply the instances (possibly through dependency injection) and using a default implementation of the factory for assembly scanning.

**Comment by FObermaier** [comment]
Do you have access to the [NetTopologySuiteBootstrapper.Bootstrap()](https://github.com/NetTopologySuite/NetTopologySuite/blob/develop/NetTopologySuite/NetTopologySuiteBootstrapper.cs) functionality. This is especially meant for  scenarios where you don't have reflection to determine the `GeometryServiceProvider.Instance`.

You should be able to assign a customized  `NtsGeometryServices` class to `GeometryServiceProvider.Instance`.

**Comment by NightOwl888** [comment]
Thanks. That did the trick.

It does feel strange to push the burden of loading this to the consumer, though. It seems like this should be automatic if I install NetTopologySuite.

**Comment by airbreather** [comment]
> Looking at the source, there doesn't appear to be a way to inject an alternative implementation of `IGeometryServices` (which would be the ideal fix). You should really consider adding a seam through the use of an [abstract factory](http://blog.ploeh.dk/2014/05/19/di-friendly-framework/) so the end user can supply the instances (possibly through dependency injection) and using a default implementation of the factory for assembly scanning.

Actually, I think the immediate problem here **may** be that we deconstruct the provided IGeometryFactory at constructor time and then reconstruct it every time a geometry is requested.  It seems like [these](https://github.com/NetTopologySuite/NetTopologySuite/blob/af938c521caf2c5bd3ce122ecc9b73b6bf10dad0/NetTopologySuite/IO/WKTReader.cs#L69) [lines](https://github.com/NetTopologySuite/NetTopologySuite/blob/af938c521caf2c5bd3ce122ecc9b73b6bf10dad0/NetTopologySuite/IO/WKTReader.cs#L445-L446) should use the `IGeometryFactory` instance that was provided to the original constructor.

IMO, `GeometryServiceProvider.Instance` should only be used as a fallback if the user didn't provide their own instance, like in this case.

**Comment by NightOwl888** [comment]
> Actually, I think the immediate problem here may be that we deconstruct the provided IGeometryFactory at constructor time and then reconstruct it every time a geometry is requested. It seems like these two lines should use the IGeometryFactory instance that was provided to the original constructor.

That seems reasonable.

> IMO, GeometryServiceProvider.Instance should only be used as a fallback if the user didn't provide their own instance, as the user seems to be doing in this case.

I am porting spatial4n from Java to support Lucene.Net which is also ported from Java, so I am a bit ignorant as to how this all works. All I know is that in Java, there is no similar call to NetTopologySuiteBootstrapper.Bootstrap() and no need to provide a custom instance in order to make the tests pass. Since it "just works" in Java, it should also "just work" in .NET (whether that be from scanning for assembly references, or calling the static NetTopologySuiteBootstrapper.Bootstrap() method internally from a static constructor of a class that is guaranteed to be instantiated when the factory is needed).

The way you have done it, the end user could still set GeometryServiceProvider.Instance, which would always occur after NetTopologySuiteBootstrapper.Bootstrap() is called (assuming it is called from the static constructor of GeometryServiceProvider), therefore it is possible for them to override the default instance.

**Comment by FObermaier** [comment]
@airbreather : WKTReader need some work as it currently does not handle`Z`, `M`, `ZM` tokens, let alone a possible definition of a spatial reference id (e.g.: `SRID=1234:POINT ZM(10 10 0 4)`

In this case, with using a provided geometry factory, you might not get a geometry having the expected srid, as that value is stored with the factory. Using a provided factory as a template and creating new factories with "correct" SRID values might pollute your heap with technically equal factories.

**Comment by TroySchmidt** [comment]
In the latest, beta release this is still an issue.  Also if you try to use other methods to initialize the WKBReader as you would expect you are directed that they are obsolete and cannot be used.

**Comment by airbreather** [comment]
Here's a bit of a status report on where we currently stand on the different pieces of this issue:

1. `NetTopologySuiteBootstrapper.Bootstrap()` -- specifically looking at the stack trace that opened this issue, you no longer **need** explicitly call this method.  We now implicitly bootstrap as soon as NTS is loaded (#227).  The only reasons I can think of why someone *might* still want to call that method explicitly are:
    - You want to have NTS's service provider override a service provider that somebody else may have already set up at the time when you call the method, or
    - You want to make absolutely sure that the NTS assembly is loaded (and therefore that the bootstrap code gets run) before doing something else that invokes `GeometryServiceProvider.Instance`.
1. Deconstruct / recreate `IGeometryFactory` in the first place: we still do this.  Perhaps we should switch to using `IGeometryServices` like `WKBReader` does.
1. `WKTReader` handling `SRID` and `Z` / `M` / `ZM`: should be fixed with #265.

**Comment by airbreather** [comment]
Status update: the only remaining negative thing I can find identified this thread that's still present as of 91cd26e is the issue of deconstructing / reconstructing the `IGeometryFactory`.

Because the original issue reported on this thread should be fixed, and the above-mentioned item shouldn't be a showstopper, I'm going to close this and open a new one focused on alternatives to this approach.
